hspec-core/src/Test/Hspec/Core/Runner.hs

```diff
@@ -99,32 +116,51 @@ ensureSeed c = case configQuickCheckSeed c of
 -- | Run given spec with custom options.
 -- This is similar to `hspec`, but more flexible.
 hspecWith :: Config -> Spec -> IO ()
-hspecWith conf spec = do
-  r <- hspecWithResult conf spec
-  unless (isSuccess r) exitFailure
+hspecWith config spec = getArgs >>= readConfig config >>= doNotLeakCommandLineArgumentsToExamples . runSpec spec >>= evaluateSummary
 
+-- | `True` if the given `Summary` indicates that there were no
+-- failures, `False` otherwise.
 isSuccess :: Summary -> Bool
 isSuccess summary = summaryFailures summary == 0
 
+-- | Exit with `exitFailure` if the given `Summary` indicates that there was at
+-- least one failure.
+evaluateSummary :: Summary -> IO ()
+evaluateSummary summary = unless (isSuccess summary) exitFailure
```

Comment:
Would `exitOnFailure` be a better name?

Reply:
I'm conflicted.  On the one hand I like `exitOnFailure`, on the other hand `evaluateSummary` may be somewhat more future proof (e.g. if we need to add more stuff to it in the future than just exiting on failure).

I'll consider and make a call by tomorrow.